src/digraph-cyclicity.jl

```diff
@@ -0,0 +1,371 @@
+    stack::Vector{Int}
+    blocked::Dict{Int, Bool}
+    blockedmap::Dict{Int, Set{Int}}
+end```
+
+Composite type that regroups the information needed for Johnson's algorithm.
+
+# Arguments
+* `stack`: the stack of visited vertices
+* `blocked`: boolean for each vertex, tell whether it is blocked or not
+* `blockedmap`: tell which vertices to unblock if the key vertex is unblocked
+"""
+type JohnsonVisitor <: Visitor
+    stack::Vector{Int}
+    blocked::Vector{Bool}
+    blockedmap::Vector{Set{Int}}
+end
+
+"""
+```JohnsonVisitor(dg::DiGraph)```
+
+Constructor of the visitor, using the directed graph information.
+"""
+JohnsonVisitor(dg::DiGraph) = JohnsonVisitor(
+Vector{Int}(),
+falses(dg.vertices),
+[Set{Int}() for i in dg.vertices]
+)
+
+"""
+```unblock!(v::T, blocked::Vector{Bool}, B::Vector{Set{Int}})```
+
+Unblock the vertices recursively.
+
+# Arguments
+* `v`: the vertex to unblock
+* `blocked`: tell whether a vertex is blocked or not
+* `B`: the map that tells if the unblocking of one vertex should unblock other vertices
+"""
+function unblock!(v::Int, blocked::Vector{Bool}, B::Vector{Set{Int}})
+    blocked[v] = false
+    for w in B[v]
+        delete!(B[v], w)
+        blocked[w] && unblock!(w, blocked, B)
+    end
+end
+
+"""
+```circuit(v::Int, dg::DiGraph, vis::JohnsonVisitor, 
+allcycles::Vector{Vector{Int}}, vmap:: Vector{Int}, startnode = v)```
+
+One step of the recursive version of simple cycle detection, using a DFS algorithm.
+
+The CIRCUIT function from [Johnson's algorithm](http://epubs.siam.org/doi/abs/10.1137/0204007), 
+recursive version. Modify the vector of cycles, when needed.
+
+# Arguments
+* `v`: the vertex considered in this iteration of the DFS
+* `dg`: the digraph from which cycles are computed
+* `visitor`: Informations needed for the cycle computation, contains:
+    * `stack`: the stack of parent vertices
+    * `blocked`: tells whether a vertex has already been explored or not
+    * `blockedmap`: mapping of the blocking / unblocking consequences
+* `allcycles`: output containing the cycles already detected
+* `vmap`: vector map containing the link from the old to the new nodes of the directed graph
+* `startnode = v`: optional argument giving the starting node. In the first iteration,
+ the same as v, otherwise it should be passed.
+
+# Returns
+* `done`: tells whether a circuit has been found in the current exploration.
+"""
+function circuit(v::Int, dg::DiGraph, vis::JohnsonVisitor, allcycles::Vector{Vector{Int}}, vmap::Vector{Int}, startnode = v)
+    done = false
+    push!(vis.stack, v)
+    vis.blocked[v] = true
+    for w in dg.fadjlist[v]
+        if w == startnode
+            push!(allcycles, vmap[vis.stack])           
+            done = true
+        elseif !vis.blocked[w]
+            circuit(w, dg, vis, allcycles, vmap, startnode) && (done = true)
+        end
+    end
+    if done
+        unblock!(v, vis.blocked, vis.blockedmap)
+    else
+        for w in dg.fadjlist[v]
```

Comment:
do not use internals like this. We expose forward and backward adjacencies via `fadj` and `badj`, respectively.